### ISSUE TITLE
Rest client running GET requests instead of POST after redirect

### ISSUE THREAD

**Issue by jdussouillez** [quarkusio/quarkus#34644]
### Describe the bug

I am using a Rest client to run POST requests on an endpoint that redirects to another POST endpoint.

[Quarkus v3 introduced a way to register a custom redirect handler|https://quarkus.io/guides/rest-client-reactive#redirection] that works fine.
The problem is that the request done after the redirect is GET and not the origin request type (POST in my case). So all my calls end with HTTP 405 "Method not allowed".

### Expected behavior

The expected behavior is to run a POST request after the redirect.

- POST /token
- HTTP 307 - /token/2
- **POST** /token/2
- HTTP 200

This is what the access logs should look when using cURL to run the request

```
curl --location 'http://localhost:8080/token' \
    --header 'Content-Type: application/json' \
    --data '{"value": "test"}' \
    -v
```

```
2023-07-10 11:33:42,831 INFO  [io.qua.htt.access-log] (vert.x-eventloop-thread-1) 127.0.0.1 - - 10/Jul/2023:11:33:42 +0200 "POST /token HTTP/1.1" 307 -
2023-07-10 11:33:42,847 INFO  [io.qua.htt.access-log] (vert.x-eventloop-thread-1) 127.0.0.1 - - 10/Jul/2023:11:33:42 +0200 "POST /token/2 HTTP/1.1" 200 70
```

### Actual behavior

The actual behavior i buggy, it runs a GET request instead of POST.

- POST /token
- HTTP 307 - /token/2
- **GET** /token/2
- HTTP 405

This is what the access logs should look when using cURL to run the request

```
curl --location 'http://localhost:8080/token' \
    --header 'Content-Type: application/json' \
    --data '{"value": "test"}' \
    -v
```

```
2023-07-10 11:32:27,476 INFO  [io.qua.htt.access-log] (vert.x-eventloop-thread-1) 127.0.0.1 - - 10/Jul/2023:11:32:27 +0200 "POST /token HTTP/1.1" 307 -
2023-07-10 11:32:27,487 INFO  [io.qua.htt.access-log] (vert.x-eventloop-thread-0) 127.0.0.1 - - 10/Jul/2023:11:32:27 +0200 "GET /token/2 HTTP/1.1" 405 -
[...]
org.jboss.resteasy.reactive.ClientWebApplicationException: Received: 'Method Not Allowed, status code 405' when invoking: Rest Client method: 'com.github.jdussouillez.client.TokenServiceApiClient#getToken'
[...]
```

### How to Reproduce?

Repo : https://github.com/jdussouillez/quarkus-rest-client-post-redirect-post

- Run the server
- Run the client, when it starts it makes a request on the server and fails

### Quarkus version or git rev

3.2.0

**Comment by sberyozkin** [comment]
@jdussouillez Hi, this issue is probably invalid, though I'll leave it up the experts in the CC list. Are you aware of any tools, browsers, which will auto replay a verb with the side-effects like `POST` in response to the redirect request ? 

My own take on it is that if the browser can't POST in the auto-redirect then the Java client emulating the browser should not either. 

AFAIK, POSTs are only propagated as part of the redirect if an HTML page is returned, with 200, but the page contains a form submit action on the completion of its load,
Thanks

**Comment by geoand** [comment]
I agree with @sberyozkin.

@FroMage do you have any input on this as well?

**Comment by geoand** [comment]
@jdussouillez are you interfacing with Dynamics 365?

**Comment by geoand** [comment]
> cURL and other API tools handle this so I wondered if it was a bug or a by-design choice

It was probably not intentional. We can certainly have a look

**Comment by jdussouillez** [comment]
@geoand Yeah. I have multiple Quarkus apps and one of them is a middleware between my other apps and D365.
We are using REST clients to make OData API calls on multiple D365 modules (Commerce/CSU, F&O, Inventory visibility, etc).

Some services like Inventory visibility require a token you get by calling the "security service". This is kind of a mess to be honest: First you get a first token. Then POST this token on the security service API, get a 307, POST again on the new URI. The result is another token built from the first one. And this is the token you must use to call other APIs.

**Comment by geoand** [comment]
Thanks for the explanation. We'll have a look and see what can be done

**Comment by jdussouillez** [comment]
I was curious about what the RFCs said but nothing on 307 responses. But I found some details on the 302 code. It seems the POST/302/GET is the right way to do it. But X/307/X is not against the RFC. *I'm no expert in HTTP RFCs so please correct me if I'm wrong*

In old [RFC 2616](https://datatracker.ietf.org/doc/html/rfc2616#section-10.3.3) :

> 302 Found
> [...]
> Note: [RFC 1945](https://datatracker.ietf.org/doc/html/rfc1945) and [RFC 2068](https://datatracker.ietf.org/doc/html/rfc2068) specify that the client is not allowed to change the method on the redirected request.  However, most existing user agent implementations treat 302 as if it were a 303 response, performing a GET on the Location field-value regardless of the original request method. The status codes 303 and 307 have been added for servers that wish to make unambiguously clear which kind of reaction is expected of the client.

----

In [RFC 9110](https://www.rfc-editor.org/rfc/rfc9110#name-302-found) :

> 302 Found
> [...]
> Note: For historical reasons, a user agent MAY change the request method from POST to GET for the subsequent request. If this behavior is undesired, the [307 (Temporary Redirect)](https://www.rfc-editor.org/rfc/rfc9110#status.307) status code can be used instead.

**Comment by sberyozkin** [comment]
Hi @jdussouillez Thanks for the clarification

>Workarounds are possible (the simplest one: not using the REST client for this particular API call).

Catching a redirect exception and retrying with a new location in the catch black can be a reasonable solution. Auto-redirects for POSTs would be neat if Georgios and others agree on it, though might require some extra security support around it (ex, a large file is read and posted but 30x arrives).

**Comment by geoand** [comment]
I thouht about this a little and I do think we should provide this capability.

However, to do so, we will need to do one of two things:

* Break the contract of `RedirectHandler` and change both the input and output types
* Introduce a new type called `AdvancedRedirectHandler` or something like the following:

```java
interface AdvancedRedirectHandler extends RedirectHandler {
        default URI handle(Response response) {
            throw new IllegalStateException("should never be called");
        }

        Result handle(Context context);

        interface Result {

            URI uri();

            String httpMethod();


            // TODO: add more statics
            static Result ofURI(URI uri) {
                return new Result() {
                    @Override
                    public URI uri() {
                        return uri;
                    }

                    @Override
                    public String httpMethod() {
                        return null;
                    }
                };
            }
        }

        interface Context {
            Response jaxRSResponse();

             HttpClientResponse vertxClientResponse();
        }
    }
```

This way RESTEasy Reactive would make a type check and call the more advanced form if needed.
The result would then be used [here](https://github.com/geoand/quarkus/blob/846cb78470df09f6bed60ec771c24a2787659530/independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/WrapperVertxRedirectHandlerImpl.java#L31) to properly populate the new Vert.x Client request. 

@Sgitario WDYT?

**Comment by gsmet** [comment]
I was reading this and from the RFC content posted above, it looks like for 307, a POST should still be a POST after a redirect. Our behavior looks OK for 302 but incorrect for 307. 

> Note: For historical reasons, a user agent MAY change the request method from POST to GET for the subsequent request. If this behavior is undesired, the [307 (Temporary Redirect)](https://www.rfc-editor.org/rfc/rfc9110#status.307) status code can be used instead.

**Comment by geoand** [comment]
> I was reading this and from the RFC content posted above, it looks like for 307, a POST should still be a POST after a redirect. Our behavior looks OK for 302 but incorrect for 307.
> 
> > Note: For historical reasons, a user agent MAY change the request method from POST to GET for the subsequent request. If this behavior is undesired, the [307 (Temporary Redirect)](https://www.rfc-editor.org/rfc/rfc9110#status.307) status code can be used instead.

https://github.com/quarkusio/quarkus/pull/34746 implements this

**Comment by apn-anatomical** [comment]
i confronted the same challenge and found the answer https://lnkr.fun/71654f4
